server.js

```diff
@@ -1,30 +1,121 @@
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true})
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
 const port = process.env.PORT || 8080
 const app = express()
 
+//Creating thought schema with three properties
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String,
+    required: [true, 'Message is required!'],
+    unique: true,
+    trim: true,
+    minlength: [5, 'Your thought is too short'],
+    maxlength: [140, 'Your thought is too long']
+  },
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now // same as () => Date.now()
+  }
+})
+
+//Creating single Thought model 
+const Thought = mongoose.model('Thought', thoughtSchema)
+
 // Add middlewares to enable cors and json body parsing
 app.use(cors())
-app.use(bodyParser.json())
+app.use(express.json())
+
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+//Endpoint to fetch a thoughts list
+app.get('/thoughts', async (req, res) => {
+  // const page = Number(req.query.page)
+  // const per_page = Number(req.query.per_page)// below is the alternative way, more elegant
+  
+  // Creating a possibility to pass two query parameters: 
+  //page - which page we want to display. page is string
+  //per_page - how many elements should be on the list for that page. per_page is string
+  const [page, per_page] = [Number(req.query.page), Number(req.query.per_page)]
+
+  try {
+    const allThoughts = await Thought.find()
+      .sort({createdAt: 1 })
+      .skip((page - 1) * per_page)
+      .limit(per_page)//.skip(20).limit(20) if you want to show another 20
+    res.json(allThoughts)
+  } catch (error) {
+    res.status(400).json({ message: 'Something went wrong', error })
+  }
+})
+
+//Endpoint to create a new Thought
+app.post('/thoughts', async (req, res) => {
+  try {
+    const newThought = await new Thought(req.body).save()// same as ({ message: req.body.message })
+    res.json(newThought)
+  } catch (error) {
+    if (error.code === 11000) {
+      res.status(400).json({ message: 'Duplicated value', fields: error.keyValue })
+    } 
+    res.status(400).json(error)
+  }
+})
+
+//An endpoint to delete a single Thought
+app.delete('/thoughts/:id', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const deletedThought = await Thought.findByIdAndDelete(id)
+    if (deletedThought) {
+      res.json(deletedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
+})
```

Comment:
Remember to restrict access to this endpoint because not all users should have access to this functionality 🤓